### ISSUE TITLE
Provide option to incorporate directory of Markdown files into generated HTML documentation

### ISSUE THREAD

**Issue by mattt** [SwiftDocOrg/swift-doc#58]
As suggested by @MaxDesiatov in #57:

> It would be great if there was an option not just to include README.md, but a given directory hierarchy of multiple markdown files that provide extended documentation: tutorials, guidelines, license, code of conduct etc.

Scoping the feature request a bit, I think a good first step would be to add a new, optional parameter to the `generate` command to specify a directory of Markdown files (`Documentation` by default) to be rendered alongside the generated symbol reference docs.

For example, given the following project structure:

```
.
├── Documentation
│   ├── Advanced-Usage.md
│   ├── FAQ.md
│   └── Getting-Started.md
└── Sources
```

`swift doc generate --format html` would produce three "Guide" pages for "Advanced Usage", "FAQ", and "Getting Started".

My only hesitation here is that it gets us into the business of static site generation more generally (a corollary of [Zawinski's Law](https://en.wikipedia.org/wiki/Jamie_Zawinski#Principles). To that end, I'm unsure on whether to should support any form of YAML front matter...

**Comment by MaxDesiatov** [comment]
Another option could be to explicitly state that `swift-doc` focuses only on documentation generation and static site generators should handle the rest if needed, which could help in avoiding feature creep. If that's the case, it would be great to have some integration between `swift-doc` and say [Publish](https://github.com/johnsundell/Publish) and maybe an ability to reuse themes across both.

**Comment by mattt** [comment]
@MaxDesiatov Please see my edit, which came in after your comment. 

I fully acknowledge the role for dedicated static site generators, and am committed to tailoring the CommonMark / SQLite output to support that kind of extensibility. That said, I think we can be smart about offering 80% of SSG functionality without blowing up the scope of our project.

**Comment by kean** [comment]
I agree with @MaxDesiatov, hosting guides alongside the API reference and using the same design language would be a major improvement in user experience.

I understand this probably blows up the scope significantly. But what I'm personally looking for is a comprehensive solution offering an experience similar or better to what we have at [developer.apple.com](developer.apple.com). This includes guides with auto-generated links for symbols (belonging to the current module and ideally also to native frameworks).

What I would also like to see is auto-generated "See Also" section (pun intended):
 
<img width="800" alt="Screen Shot 2020-04-08 at 23 51 54" src="https://user-images.githubusercontent.com/1567433/78856231-2c463080-79f4-11ea-8b9f-30352e862a5d.png">

**Comment by kean** [comment]
Hey, I hacked together a quick prototype using SwiftDoc and some of its dependencies. I hosted the end result [here](https://kean-org.github.io/docs/nuke-docs-beta/guides/image-pipeline.html).

<img width="1375" alt="Screen Shot 2020-04-10 at 20 39 16" src="https://user-images.githubusercontent.com/1567433/79031244-6f76df80-7b6b-11ea-9d37-876953961050.png">

Some of the issues I encountered:

- The [approach](https://github.com/SwiftDocOrg/swift-doc/pull/65) with `<base href>` completely breaks anchors (e.g. `#deduplication`). They now all redirect to the root URL instead of opening an anchor in the current document.
- There are limitations with the current approach of linking code elements. Running a highlighter and then attempting to add links only gets you so far. For example, it doesn't work for nested types, e.g. `ImageProcessors.Resize`.
- There are some severe limitations in [Markup](https://github.com/SwiftDocOrg/Markup.git), [SwiftSyntaxHighlighter](https://github.com/NSHipster/SwiftSyntaxHighlighter.git) is also not doing a great job with some code samples.
- Already visited links are grayed-out

<img width="180" alt="Screen Shot 2020-04-10 at 20 42 56" src="https://user-images.githubusercontent.com/1567433/79031309-f2983580-7b6b-11ea-80d3-d690086b1338.png">

I also started wondering what the user experience for these kinds of guides should be. Ideally, I would probably like it to be similar to that of Xcode Playgrounds, except that instead of jump-to-definition, clicking on an element would open an API reference. It sounds a lot like a pipe-dream, to be honest. But maybe this is something that could be implemented using the existing tools. I'm not familiar with [Swift LSP](https://github.com/apple/sourcekit-lsp) yet, but it might be something worth looking into.

**Comment by Lukas-Stuehrk** [comment]
I worked on a small proof of concept to continue the discussion: https://github.com/SwiftDocOrg/swift-doc/compare/master...Lukas-Stuehrk:AdditionalDocumentation

## Problems solved:
- You can provide  additional guides with a new flag`--documentation path/to/the/directory`.
- Providing a title: I simply take the title from the first level 1 heading in the document. If the document does not provide such a heading, I take the filename. I somehow like this approach. No processing instructions or any other mechanism needed.

## Things to be discussed

- I simply added a new section in the homepage and the sidebar and called it "Additional Guides" (very inspired from jazzy).
- It's possible to overwrite a generated page (e.g. for a type) with a non-generated guide. I'm not sure if this is a cool feature or by accident because of my implementation approach.
- I flatten the file structure of the additional documents and do not preserve the tree structure of their folder. This might be very surprising, especially if you link between documents. This is rather easy to fix, I simply did not want too much time on the routing, especially because @mattt already has a WIP pull request where the routing implementation is changed.
- This adds again a lot of functionality into `swift-doc` (the executable) and not to `SwiftDoc` (the library). Currently we can't add tests to `swift-doc`, except for the end-to-end tests. This starts to not feel right anymore, I think there are some parts that could be tested.

## The ugly parts:
- The HTML and Markdown generation is....let's say it's not very sophisticated.

Please let me know what you think. I'm happy to work on this and to bring it into a mergeable state.
